### PR TITLE
Order fields/properties better based on existing class structure.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/19239

This tweaks "generate field/prop" in a few ways.  First, when generating a field from an assignment, we look to see if there are surrounding assignments in the previous/next statement to another field.  If so, and that other assignment is to a readonly field, we suggest generating a readonly field above a writable field.  The idea here being that users will generally be initializing a set of fields that are similar in terms of if they are readonly or not.

Second, if the user is creating a property from within a constructor, then we prefer generating an auto-prop over a property with a private setter.  Previously, the private setter approach was the closest we could come to a readonly property.  But now that we have autoprops, those can be preferred.

Third, we use the surrounding assignments to help inform where to place the field/property.  i.e. if you add an assignment after an existing assignment to a field, we'll attempt to place the new field after the field that was assigned in previous statement.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz 


**Customer scenario**

This adds several improvements to how we generate fields/properties in the IDE that should generally make users' lives more pleasant (ideally without them even realizing that anything has changed).  When generating things previously we used some heuristics to order and place things to try to match what we thought the user's intent was.  We now continue that even further, making it so the top suggestion we provide is more likely teh right one.  For example, if the user is writing a sequence of readonly fields, we'll then prefer to create a readonly field over a writable one.  Similarly, we'll use the surrounding context to better determine where the user wants to place these members (so that tehy don't have to go manually order htings themselves).

**Bugs this fixes:**

#19239
https://github.com/dotnet/roslyn/issues/18988

**Workarounds, if any**

None.

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

This is the infinite tail of doing a better job making features feel like they're super intelligent.

**How was the bug found?**

Customer reports.

File: src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs [hunk 1]
```diff
@@ -252,9 +258,131 @@ internal bool CanGenerateLocal()
...
+                var simpleName = this.SimpleNameOpt;
+
+                if (syntaxFacts.IsLeftSideOfAssignment(simpleName))
+                {
+                    var assignmentStatement = simpleName.Ancestors().FirstOrDefault(syntaxFacts.IsSimpleAssignmentStatement);
+                    if (assignmentStatement != null)
+                    {
+                        syntaxFacts.GetPartsOfAssignmentStatement(assignmentStatement, out var left, out var right);
+
+                        if (left == simpleName)
```
Comment (Pilchie) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs]: Why?  What about `this.name` etc?  This check doesn't seem necessary.

File: src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs [hunk 2]
```diff
@@ -252,9 +258,131 @@ internal bool CanGenerateLocal()
...
+                            return true;
+                        }
+
+                        return false;
+                }
+
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            private int GetStatementindex(ChildSyntaxList children, SyntaxNode statement)
```
Comment (Pilchie) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs]: Spelling - capitalize I

File: src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs [hunk 3]
```diff
@@ -252,9 +258,131 @@ internal bool CanGenerateLocal()
...
+                {
+                    case IFieldSymbol field: return field.IsReadOnly;
+                    case IPropertySymbol property:
+                        if (property.SetMethod == null)
+                        {
+                            return true;
+                        }
+
+                        if (property.ContainingType.GetMembers().OfType<IFieldSymbol>().Any(
+                                f => property.SetMethod.Equals(f.AssociatedSymbol)))
```
Comment (Pilchie) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs]: I'm having a really hard time decyphering what this expression is trying to represent.

"If this type has a field associated with this property's set method"?  Why does that make it readonly?

File: src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs (hunk 2 above)
Comment (Pilchie) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs]: Also - I think we have a faster search that binary searches by position of the node?

Comment (Pilchie) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs]: I'm thinking about some of the multi-thousand line `InitializeComponent` methods I've seen.

File: src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs]: Fixed.

**Pull request merged**